### PR TITLE
feat(limitTo): add support for array-likes

### PR DESCRIPTION
**Pull request by wesleycho**
- Add support for array-like objects

This implements #14657.

### CODE REVIEW

File: test/ng/filter/limitToSpec.js [hunk 1]
```diff
@@ -4,15 +4,35 @@ describe('Filter: limitTo', function() {
...
+      6: 'g',
+      7: 'h',
+      get length() {
+        return Object.keys(this).length - 1;
+      }
+    };
     limitTo = $filter('limitTo');
   }));
 
+  function isInObject(obj) {
```
Comment (campersau) [inline test/ng/filter/limitToSpec.js]: This is never used.

File: src/ng/filter/limitTo.js [hunk 2]
```diff
@@ -108,19 +108,25 @@ function limitToFilter() {
...
       } else {
-        return input.slice(Math.max(0, begin + limit), begin);
+        return sliceFn(input, Math.max(0, begin + limit), begin);
       }
     }
   };
 }
+
+function sliceFn(input, idx, length) {
+  if (isString(input)) return input.slice(idx, length);
```
Comment (gkalpak) [inline src/ng/filter/limitTo.js]: I would check if `input` has a `slice` method (or property) so we support Array-like objects with custon `slice` methods.

File: src/ng/filter/limitTo.js [hunk 3]
```diff
@@ -108,19 +108,25 @@ function limitToFilter() {
...
+
+  return Array.prototype.slice.call(input, idx, length);
```
Comment (gkalpak) [inline src/ng/filter/limitTo.js]: You can also call the "IIFE global" `slice` method: `slice.call(...)`
(It is initialized as `[].slice`.)

File: test/ng/filter/limitToSpec.js [hunk 4]
```diff
@@ -71,6 +105,11 @@ describe('Filter: limitTo', function() {
     expect(limitTo(items, 'undefined')).toEqual(items);
     expect(limitTo(items, null)).toEqual(items);
     expect(limitTo(items, undefined)).toEqual(items);
+    expect(limitTo(arrayLike, 'bogus')).toEqual(arrayLike);
+    expect(limitTo(arrayLike, 'null')).toEqual(arrayLike);
+    expect(limitTo(arrayLike, 'undefined')).toEqual(arrayLike);
+    expect(limitTo(arrayLike, null)).toEqual(arrayLike);
+    expect(limitTo(arrayLike, undefined)).toEqual(arrayLike);
```
Comment (gkalpak) [inline test/ng/filter/limitToSpec.js]: I would rather use `toBe` in this test as it captures the intent better. (I know it was `toEqual` even before this PR.)

File: test/ng/filter/limitToSpec.js [hunk 5]
```diff
@@ -97,12 +136,16 @@ describe('Filter: limitTo', function() {
     expect(limitTo(str, '3', 'undefined')).toEqual(limitTo(str, '3'));
     expect(limitTo(str, '-3', null)).toEqual(limitTo(str, '-3', 0));
     expect(limitTo(str, 3, undefined)).toEqual(limitTo(str, 3));
+    expect(limitTo(arrayLike, 3, 'bogus')).toEqual(limitTo(items, 3, 0));
+    expect(limitTo(arrayLike, -3, 'null')).toEqual(limitTo(items, -3));
+    expect(limitTo(arrayLike, '3', 'undefined')).toEqual(limitTo(items, '3', 0));
+    expect(limitTo(arrayLike, '-3', null)).toEqual(limitTo(items, '-3'));
+    expect(limitTo(arrayLike, 3, undefined)).toEqual(limitTo(items, 3, 0));
```
Comment (gkalpak) [inline test/ng/filter/limitToSpec.js]: _items_  -->  _arrayLike_

File: test/ng/filter/limitToSpec.js [hunk 6]
```diff
@@ -97,12 +136,16 @@ describe('Filter: limitTo', function() {
...
   });
 
-  it('should return input if not String or Array or Number', function() {
+  it('should return input if not String or array-like or Number', function() {
```
Comment (gkalpak) [inline test/ng/filter/limitToSpec.js]: Remove _"String or"_.

File: test/ng/filter/limitToSpec.js [hunk 7]
```diff
@@ -97,12 +136,16 @@ describe('Filter: limitTo', function() {
...
     expect(limitTo(null, 1)).toEqual(null);
     expect(limitTo(undefined, 1)).toEqual(undefined);
-    expect(limitTo({}, 1)).toEqual({});
```
Comment (gkalpak) [inline test/ng/filter/limitToSpec.js]: Why did you remove this ?

File: test/ng/filter/limitToSpec.js [hunk 8]
```diff
@@ -111,8 +154,13 @@ describe('Filter: limitTo', function() {
     expect(limitTo(items, '9')).toEqual(items);
     expect(limitTo(items, -9)).toEqual(items);
     expect(limitTo(items, '-9')).toEqual(items);
+    expect(limitTo(arrayLike, 9)).toEqual(items);
+    expect(limitTo(arrayLike, '9')).toEqual(items);
+    expect(limitTo(arrayLike, -9)).toEqual(items);
+    expect(limitTo(arrayLike, '-9')).toEqual(items);
 
-    expect(limitTo(items, 9)).not.toBe(items);
```
Comment (gkalpak) [inline test/ng/filter/limitToSpec.js]: Why did you remove this ?

File: test/ng/filter/limitToSpec.js [hunk 9]
```diff
@@ -111,8 +154,13 @@ describe('Filter: limitTo', function() {
     expect(limitTo(items, '9')).toEqual(items);
     expect(limitTo(items, -9)).toEqual(items);
     expect(limitTo(items, '-9')).toEqual(items);
+    expect(limitTo(arrayLike, 9)).toEqual(items);
+    expect(limitTo(arrayLike, '9')).toEqual(items);
+    expect(limitTo(arrayLike, -9)).toEqual(items);
+    expect(limitTo(arrayLike, '-9')).toEqual(items);
```
Comment (gkalpak) [inline test/ng/filter/limitToSpec.js]: _items_  -->  _arrayLike_

File: test/ng/filter/limitToSpec.js [hunk 10]
```diff
@@ -111,8 +154,13 @@ describe('Filter: limitTo', function() {
...
 
-    expect(limitTo(items, 9)).not.toBe(items);
+    expect(limitTo(arrayLike, 9)).not.toBe(items);
+    expect(limitTo(arrayLike, 9)).not.toBe(items);
```
Comment (gkalpak) [inline test/ng/filter/limitToSpec.js]: _items_  -->  _arrayLike_

File: test/ng/filter/limitToSpec.js [hunk 11]
```diff
@@ -129,6 +177,10 @@ describe('Filter: limitTo', function() {
     expect(limitTo(items, 'Infinity')).toEqual(items);
     expect(limitTo(items, -Infinity)).toEqual(items);
     expect(limitTo(items, '-Infinity')).toEqual(items);
+    expect(limitTo(arrayLike, Infinity)).toEqual(items);
+    expect(limitTo(arrayLike, 'Infinity')).toEqual(items);
+    expect(limitTo(arrayLike, -Infinity)).toEqual(items);
+    expect(limitTo(arrayLike, '-Infinity')).toEqual(items);
```
Comment (gkalpak) [inline test/ng/filter/limitToSpec.js]: _items_  -->  _arrayLike_

**Comment by gkalpak** [comment]
Generally looks good - I left somme comments.
I think it would be nice to test against a couple more Array-likes (e.g. a NodeList, a jqLite/jQuery collection).

File: src/ng/filter/limitTo.js (hunk 2 above)
Comment (wesleycho) [inline src/ng/filter/limitTo.js]: This separation is specifically because `String.prototype.slice` returns a string, while `Array.prototype.slice` returns an array.

**Comment by wesleycho** [comment]
Made adjustments as per comments.

I do have a question though - I notice in the `orderBy` filter, it only checks against a similar object I have constructed here, and not against typical alternative array-likes. Does it make sense to test those here given that?

File: src/ng/filter/limitTo.js (hunk 2 above)
Comment (gkalpak) [inline src/ng/filter/limitTo.js]: OK, let's stick with the current separation, so we are always returnign either a string or an array.

File: test/ng/filter/limitToSpec.js [hunk 12]
```diff
@@ -44,38 +64,52 @@ describe('Filter: limitTo', function() {
...
+    expect(limitTo(arrayLike, 'null')).toBe(arrayLike);
+    expect(limitTo(arrayLike, 'undefined')).toBe(arrayLike);
+    expect(limitTo(arrayLike, null)).toBe(arrayLike);
+    expect(limitTo(arrayLike, undefined)).toBe(arrayLike);
   });
 
   it('should return an empty string when X = 0', function() {
     expect(limitTo(str, 0)).toEqual("");
     expect(limitTo(str, '0')).toEqual("");
+    expect(limitTo({}, 1)).toEqual({});
```
Comment (gkalpak) [inline test/ng/filter/limitToSpec.js]: Why this ?

File: test/ng/filter/limitToSpec.js [hunk 13]
```diff
@@ -111,8 +149,13 @@ describe('Filter: limitTo', function() {
     expect(limitTo(items, '9')).toEqual(items);
     expect(limitTo(items, -9)).toEqual(items);
     expect(limitTo(items, '-9')).toEqual(items);
+    expect(limitTo(arrayLike, 9)).toEqual(items);
+    expect(limitTo(arrayLike, '9')).toEqual(items);
+    expect(limitTo(arrayLike, -9)).toEqual(items);
+    expect(limitTo(arrayLike, '-9')).toEqual(items);
 
     expect(limitTo(items, 9)).not.toBe(items);
+    expect(limitTo(arrayLike, 9)).not.toBe(items);
```
Comment (gkalpak) [inline test/ng/filter/limitToSpec.js]: Shouldn't this be `arrayLike` (instead of `items`) ?

**Comment by gkalpak** [comment]
`$filter` has a couple of array-likes (see [filterSpec.js#L428-L443][1]), but I don't think consistency is important here.

I don't mean add all array-likes in all usecases, just add a simple testcase that tests basic limiting on 2-3 array-like types. But I don't feel strongly about it, so feel free to ignore :smile: 

[1]: https://github.com/angular/angular.js/blob/cfc8b41771e6f2f5bcf692400e960144009b1f18/test/ng/filter/filterSpec.js#L428-L443

**Comment by wesleycho** [comment]
Updated as per comments (sloppiest PR I've done in a while...) - I also added a simple test case for basic array likes at the very end.

File: src/ng/filter/limitTo.js (hunk 2 above)
Comment (wesleycho) [inline src/ng/filter/limitTo.js]: One alternative implementation I can do is instead of the `isString(input)` check, I can do
```js
return input.slice ? input.slice.call(input, idx, length) : slice.call(input, idx, length);
```
Thoughts?

Comment (wesleycho) [inline src/ng/filter/limitTo.js]: Nevermind, looks like this fails on falsy types.

Comment (gkalpak) [inline src/ng/filter/limitTo.js]: I think it's best to keep things as is, so we don't break the guarantee that `limitTo` returns an Array or String.
(I don't see any real benefit in breaking that contract.)

**Comment by wesleycho** [comment]
Bah, I did a mistaken force push to try to restart the CI - will push again when I'm at work tomorrow, some changes got wiped out.

**Comment by wesleycho** [comment]
This should be up to date now.

File: src/ng/filter/limitTo.js [hunk 14]
```diff
@@ -108,19 +108,25 @@ function limitToFilter() {
...
+        return sliceFn(input, limit, input.length);
       } else {
-        return input.slice(Math.max(0, begin + limit), begin);
+        return sliceFn(input, Math.max(0, begin + limit), begin);
       }
     }
   };
 }
+
+function sliceFn(input, idx, length) {
```
Comment (gkalpak) [inline src/ng/filter/limitTo.js]: Nitpick: Renaming the arguments to `input, begin, end` sounds more intuitive.

File: test/ng/filter/limitToSpec.js [hunk 15]
```diff
@@ -97,12 +130,16 @@ describe('Filter: limitTo', function() {
...
+    expect(limitTo(arrayLike, '3', 'undefined')).toEqual(limitTo(arrayLike, '3', 0));
+    expect(limitTo(arrayLike, '-3', null)).toEqual(limitTo(arrayLike, '-3'));
+    expect(limitTo(arrayLike, 3, undefined)).toEqual(limitTo(arrayLike, 3, 0));
   });
 
-  it('should return input if not String or Array or Number', function() {
+  it('should return input if not array-like or Number', function() {
     expect(limitTo(null, 1)).toEqual(null);
     expect(limitTo(undefined, 1)).toEqual(undefined);
-    expect(limitTo({}, 1)).toEqual({});
```
Comment (gkalpak) [inline test/ng/filter/limitToSpec.js]: Why that ?

File: test/ng/filter/limitToSpec.js [hunk 16]
```diff
@@ -153,19 +201,38 @@ describe('Filter: limitTo', function() {
...
+    function getArguments() {
+      return arguments;
+    }
+    var argsObj = getArguments({name: 'Misko'}, {name: 'Igor'}, {name: 'Brad'});
+
+    var nodeList = jqLite("<p><span>Misko</span><span>Igor</span><span>Brad</span></p>")[0].childNodes;
+
+    expect(limitTo(argsObj, 2).length).toBe(2);
+    expect(limitTo('abc', 1).length).toBe(1);
+    expect(limitTo(nodeList, 3).length).toBe(3);
```
Comment (gkalpak) [inline test/ng/filter/limitToSpec.js]: This is not testing anything. You need a `limit` that is less than the actual length.

**Comment by gkalpak** [comment]
Now it's time for the "real stuff": We need to update the docs :smiley:

**Comment by wesleycho** [comment]
Added documentation update, might need tweaking on wording (or maybe removing the bit about strings or numbers).

**Comment by gkalpak** [comment]
@wesleycho, I don't see any docs :worried:

**Comment by wesleycho** [comment]
That's because I'm a moron :( - pushed

File: src/ng/filter/limitTo.js [hunk 17]
```diff
@@ -9,9 +9,10 @@
  * Creates a new array or string containing only a specified number of elements. The elements
  * are taken from either the beginning or the end of the source array, string or number, as specified by
  * the value and sign (positive or negative) of `limit`. If a number is used as input, it is
- * converted to a string.
+ * converted to a string. Array-like values (e.g. NodeLists, jQuery objects, TypedArrays, Strings, etc)
+ * are also supported.
  *
- * @param {Array|string|number} input Source array, string or number to be limited.
+ * @param {Array} input Array or array-like to be limited.
```
Comment (petebacondarwin) [inline src/ng/filter/limitTo.js]: So the param type should be `{Array|string|number|object}` right?

Comment (gkalpak) [inline src/ng/filter/limitTo.js]: In `orderBy`, we use `ArrayLike` (which is totally made-up, but pretty descriptive).

Comment (petebacondarwin) [inline src/ng/filter/limitTo.js]: OK. That works for me - does `orderBy` convert numbers to strings too?

Comment (gkalpak) [inline src/ng/filter/limitTo.js]: `orderBy` only accepts `{Array|ArrayLike}` (and calls `Array.prototype.map` on its input).

**Comment by gkalpak** [comment]
I just realized, @googlebot didn't bother verifying the CLA fos us :unamused:
Can we trigger it manually ?

**Comment by petebacondarwin** [comment]
Not sure but @wesleycho is definitely a yes already :-)

**Comment by gkalpak** [comment]
I'll add the label manually, because my PR-merging script doesn't proceed otherwise :)
